### PR TITLE
WIP: Dynamic OG complex field

### PR DESCRIPTION
**Pull request by RoySegall**

### CODE REVIEW

**Comment by RoySegall** [comment]
@amitaibu I think i'm on to something here:
create_group_content___site-install

**Comment by pfrenssen** [comment]
Oh boy, looks like this will invalidate all my work on OgComplexWidgetTest :) Never mind, this looks like a great improvement!

**Comment by RoySegall** [comment]
:+1:

**Comment by RoySegall** [comment]
I was pretty amazed to get to this point. I think the multiple autocomplete widget is the bad boy in this story since we need to wrap it with add another item but i'll handle it.

**Comment by RoySegall** [comment]
I made another improvement to the API. Some minor documentation:

``` php
    dpm(OgGroupAudienceHelper::getAvailableWidgets());
    OgGroupAudienceHelper::setWidgets('node', 'group_content', 'og_group_ref', [
      'default' => 'options_select',
      'admin' => 'options_select',
    ]);
```

**Comment by amitaibu** [comment]
nice! now only to see if you can save and edit ;)

**Comment by pfrenssen** [comment]
This is actually not interfering at all with the OgComplexWidgetTest from #140, this is even answering some questions I had in there :)

The relevant part is the `testFieldModes()` test, that actually tests this functionality. I will take it out of #140 and move it here. That way this is instantly tested, and it also reduces the scope of #140.

**Comment by pfrenssen** [comment]
@RoySegall OK if I work on the test part of this? I won't touch your code.

**Comment by RoySegall** [comment]
I have a couple of things i'd like to do you could join in such as ordering the code.

I'll ping you.

**Comment by RoySegall** [comment]
I tried to do a couple of things but you can join in although I think it would be very confusing.

**Comment by pfrenssen** [comment]
OK no problem, I'm going to work on #162 today.

**Comment by RoySegall** [comment]
Didn't got the time but all the default groups selection are working. We need to handle the other group.

Also we need to handle views selection handler but we can keep for it later.

File: src/Plugin/Field/FieldWidget/OgComplex.php [hunk 1]
```diff
@@ -70,120 +72,46 @@ public function form(FieldItemListInterface $items, array &$form, FormStateInter
...
-        break;
-    }
+    $handler = OgGroupAudienceHelper::renderWidget($this->fieldDefinition, $widget_id);
 
-    $title = $this->fieldDefinition->getLabel();
-    $description = FieldFilteredMarkup::create(\Drupal::token()->replace($this->fieldDefinition->getDescription()));
+    if ($handler instanceof EntityReferenceAutocompleteWidget) {
+      // No need for extra work here since we already extending the auto
+      // complete handler.
+      return $multiple;
```
Comment (RoySegall) [inline src/Plugin/Field/FieldWidget/OgComplex.php:91]: @amitaibu ^^

File: src/OgGroupAudienceHelper.php [hunk 2]
```diff
@@ -111,4 +118,101 @@ public static function getMatchingField(ContentEntityInterface $entity, $group_t
...
+
+      if (!in_array('entity_reference', $definition['field_types'])) {
+        continue;
+      }
+
+      $widgets[] = $id;
+    }
+
+    return $widgets;
+  }
```
Comment (pfrenssen) [inline src/OgGroupAudienceHelper.php:158]: What's the reason for splitting this off as static methods in OgGroupAudienceHelper? Are you planning on reusing them somewhere else?

Comment (RoySegall) [inline src/OgGroupAudienceHelper.php:158]: Yes - we would like to get all the available widget so the user could pick them through the UI. No?

Comment (pfrenssen) [inline src/OgGroupAudienceHelper.php:158]: OK!

Comment (RoySegall) [inline src/OgGroupAudienceHelper.php:158]: Way to go! That's the spirit!

**Comment by amitaibu** [comment]
@RoySegall can you please give a quick update on what you accomplished yesterday.

**Comment by RoySegall** [comment]
Yeah. Sure. The handlers give us the correct groups - my groups and other groups. There are still some WTF with the other group widget when it's set to autocomplete and when selecting groups via autocomplete the groups are being ignored - problem with FAPI and how the values are processed.

Except that - we all set to go :)

**Comment by RoySegall** [comment]
Update: I managed to have a single auto complete element and get his selected group. The next part will be to have a multiple one but I think that in this case we would to create our own 'Add another button' as we did at the beginning. I'll update.

**Comment by RoySegall** [comment]
I pushed now a small work on the custom widget - https://gfycat.com/DisloyalRegularKodiakbear - looks good for now.

**Comment by amitaibu** [comment]
@RoySegall what's the status on this one - what's left?

**Comment by RoySegall** [comment]
I need to handle "Add another button" for the other groups widget.

**Comment by RoySegall** [comment]
Closing infavor of #248.

**Pull request closed without merging**